# E-commerce Data Exploration

This notebook explores the Olist Brazilian E-Commerce dataset to understand its structure, relationships and key variables before performing the SQL and business analysis.

The goal of this initial exploration is to understand the available data and identify the tables and fields needed to analyze sales, customers, product performance and delivery experience.

In [2]:
import pandas as pd

## Orders

In [3]:
orders = pd.read_csv("../data/raw/olist_orders_dataset.csv")

In [4]:
orders.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [5]:
orders.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='str')

### Orders table

Each row represents an e-commerce order.

- `order_id`: Unique identifier for each order.
- `customer_id`: Identifier of the customer associated with the order.
- `order_status`: Current/final status of the order.
- `order_purchase_timestamp`: Date and time when the purchase was placed.
- `order_approved_at`: Date and time when the order/payment was approved.
- `order_delivered_carrier_date`: Date and time when the order was handed over to the carrier.
- `order_delivered_customer_date`: Actual delivery date to the customer.
- `order_estimated_delivery_date`: Estimated delivery date.

These timestamps can later be used to analyze delivery times, delays and their potential relationship with customer satisfaction.

## Customers

In [6]:
customers = pd.read_csv("../data/raw/olist_customers_dataset.csv")

customers.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [7]:
customers.columns

Index(['customer_id', 'customer_unique_id', 'customer_zip_code_prefix',
       'customer_city', 'customer_state'],
      dtype='str')

In [8]:
customers["customer_id"].nunique()

99441

In [9]:
customers["customer_unique_id"].nunique()

96096

### Customers table

Each row represents a customer record associated with an order.

- `customer_id`: Identifier associated with the customer for a specific order. The same person may have different `customer_id` values across different orders.
- `customer_unique_id`: Unique identifier for the actual customer. This allows us to identify multiple orders made by the same person.
- `customer_zip_code_prefix`: Prefix of the customer's ZIP code.
- `customer_city`: Customer's city.
- `customer_state`: Customer's state in Brazil.

The distinction between `customer_id` and `customer_unique_id` is important for customer analysis. `customer_id` is used to connect the customers table with the orders table, while `customer_unique_id` should be used to identify unique and repeat customers.

This table will later be used to analyze customer distribution, repeat purchases and customer retention.

## Order Items

In [10]:
order_items = pd.read_csv("../data/raw/olist_order_items_dataset.csv")

In [11]:
order_items.head(15)

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14
5,00048cc3ae777c65dbb7d2a0634bc1ea,1,ef92defde845ab8450f9d70c526ef70f,6426d21aca402a131fc0a5d0960a3c90,2017-05-23 03:55:27,21.90,12.69
6,00054e8431b9d7675808bcb819fb4a32,1,8d4f2bb7e93e6710a28f34fa83ee7d28,7040e82f899a04d1b434b795a43b4617,2017-12-14 12:10:31,19.90,11.85
7,000576fe39319847cbb9d288c5617fa6,1,557d850972a7d6f792fd18ae1400d9b6,5996cddab893a4652a15592fb58ab8db,2018-07-10 12:30:45,810.00,70.75
8,0005a1a1728c9d785b8e2b08b904576c,1,310ae3c140ff94b03219ad0adc3c778f,a416b6a846a11724393025641d4edd5e,2018-03-26 18:31:29,145.95,11.65
9,0005f50442cb953dcd1d21e1fb923495,1,4535b0e1091c278dfd193e5a1d63b39f,ba143b05f0110f0dc71ad71b4466ce92,2018-07-06 14:10:56,53.99,11.40


In [12]:
order_items.columns

Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value'],
      dtype='str')

### Order Items table

Each row represents an individual item within an order. A single order can contain multiple items.

- `order_id`: Identifier of the order. Used to connect order items with the orders table.
- `order_item_id`: Sequential number identifying each item within an order.
- `product_id`: Identifier of the product purchased.
- `seller_id`: Identifier of the seller who sold the product.
- `shipping_limit_date`: Deadline for the seller to hand the item over for shipping.
- `price`: Price of the individual item.
- `freight_value`: Shipping cost associated with the item.

An `order_id` can appear multiple times when an order contains more than one item. Therefore, the number of rows in this table does not represent the number of orders.

This table will be used to analyze product sales, revenue, freight costs and seller performance.

## Products

In [13]:
products = pd.read_csv("../data/raw/olist_products_dataset.csv")

In [14]:
products.head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0


In [15]:
products.columns

Index(['product_id', 'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm'],
      dtype='str')

### Products table

Each row represents a product available in the marketplace.

- `product_id`: Unique identifier for each product.
- `product_category_name`: Product category in Portuguese.
- `product_name_lenght`: Number of characters in the product name.
- `product_description_lenght`: Number of characters in the product description.
- `product_photos_qty`: Number of product photos.
- `product_weight_g`: Product weight in grams.
- `product_length_cm`: Product length in centimeters.
- `product_height_cm`: Product height in centimeters.
- `product_width_cm`: Product width in centimeters.

The `product_id` connects this table with the Order Items table.

Product categories are currently in Portuguese. The `product_category_name_translation` dataset will later be used to translate category names into English.

This table will mainly be used to analyze sales and revenue by product category.

## Product Category Translation

In [16]:
product_category_translation = pd.read_csv("../data/raw/product_category_name_translation.csv")

In [17]:
product_category_translation.head()

,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto
3,cama_mesa_banho,bed_bath_table
4,moveis_decoracao,furniture_decor


In [18]:
product_category_translation.columns

Index(['product_category_name', 'product_category_name_english'], dtype='str')

### Product Category Translation table

This is a lookup table used to translate product category names from Portuguese into English.

- `product_category_name`: Original product category name in Portuguese.
- `product_category_name_english`: English translation of the product category.

The `product_category_name` field connects this table with the Products table.

This table will be used to make product category analysis and dashboard visualizations easier to understand for an international audience.

## Payments

In [19]:
payments = pd.read_csv("../data/raw/olist_order_payments_dataset.csv")

In [20]:
payments.head()

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


In [21]:
payments.columns

Index(['order_id', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value'],
      dtype='str')

### Payments table

Each row represents a payment transaction associated with an order. An order may contain more than one payment transaction.

- `order_id`: Identifier of the order. Used to connect payments with the Orders table.
- `payment_sequential`: Sequential number of the payment transaction within an order.
- `payment_type`: Payment method used by the customer.
- `payment_installments`: Number of installments selected for the payment.
- `payment_value`: Total value of the payment transaction.

The same `order_id` may appear multiple times when an order uses more than one payment transaction.

It is important to distinguish `payment_value` from the `price` field in the Order Items table. Their relationship will be investigated before defining the project's revenue metrics.

This table will be used to analyze payment methods, installments and customer payment values.

## Reviews

In [22]:
reviews = pd.read_csv("../data/raw/olist_order_reviews_dataset.csv")

In [23]:
reviews.head()

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53


In [24]:
reviews.columns

Index(['review_id', 'order_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp'],
      dtype='str')

### Reviews table

Each row represents a customer review associated with an order.

- `review_id`: Unique identifier for the review.
- `order_id`: Identifier of the order associated with the review.
- `review_score`: Customer rating, ranging from 1 to 5.
- `review_comment_title`: Optional title provided with the review.
- `review_comment_message`: Optional written feedback from the customer.
- `review_creation_date`: Date when the review survey was created.
- `review_answer_timestamp`: Date and time when the customer submitted the review.

Customers can provide a review score without leaving a written comment, so `review_score` will be the primary variable used to measure customer satisfaction.

The `order_id` connects this table with the Orders table.

This table will be used to analyze customer satisfaction and investigate whether delivery performance is associated with lower or higher review scores.

## Sellers

In [25]:
sellers = pd.read_csv("../data/raw/olist_sellers_dataset.csv")

In [26]:
sellers.head()

,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


In [27]:
sellers.columns

Index(['seller_id', 'seller_zip_code_prefix', 'seller_city', 'seller_state'], dtype='str')

### Sellers table

Each row represents a seller operating in the Olist marketplace.

- `seller_id`: Unique identifier for each seller.
- `seller_zip_code_prefix`: Prefix of the seller's ZIP code.
- `seller_city`: City where the seller is located.
- `seller_state`: Brazilian state where the seller is located.

The `seller_id` connects this table with the Order Items table.

This table will be used to analyze seller performance and the geographical distribution of sellers.

## Geolocation

In [28]:
geolocation = pd.read_csv("../data/raw/olist_geolocation_dataset.csv")

In [29]:
geolocation.head()

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP


In [30]:
geolocation.columns

Index(['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng',
       'geolocation_city', 'geolocation_state'],
      dtype='str')

### Geolocation table

Each row represents a geographical location associated with a Brazilian ZIP code prefix.

- `geolocation_zip_code_prefix`: Prefix of the Brazilian ZIP code.
- `geolocation_lat`: Latitude of the geographical location.
- `geolocation_lng`: Longitude of the geographical location.
- `geolocation_city`: City associated with the location.
- `geolocation_state`: Brazilian state associated with the location.

A ZIP code prefix can appear multiple times with slightly different coordinates. Therefore, this table should not be joined directly with customer or seller data without first handling duplicate ZIP code prefixes.

This table can later be used to support geographical analysis and map visualizations.

# Initial Data Quality Assessment

In [31]:
orders.shape

(99441, 8)

In [32]:
orders.dtypes

order_id                         str
customer_id                      str
order_status                     str
order_purchase_timestamp         str
order_approved_at                str
order_delivered_carrier_date     str
order_delivered_customer_date    str
order_estimated_delivery_date    str
dtype: object

In [33]:
orders.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [34]:
orders.duplicated().sum()

np.int64(0)

In [36]:
orders["order_id"].duplicated().sum()

np.int64(0)

In [40]:
orders[orders["order_delivered_customer_date"].isnull()]["order_status"].value_counts()

order_status
shipped        1107
canceled        619
unavailable     609
invoiced        314
processing      301
delivered         8
created           5
approved          2
Name: count, dtype: int64

In [41]:
orders[
    (orders["order_status"] == "delivered") &
    (orders["order_delivered_customer_date"].isnull())
]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
3002,2d1e2d5bf4dc7227b3bfebb81328c15f,ec05a6d8558c6455f0cbbd8a420ad34f,delivered,2017-11-28 17:44:07,2017-11-28 17:56:40,2017-11-30 18:12:23,NaN,2017-12-18 00:00:00
20618,f5dd62b788049ad9fc0526e3ad11a097,5e89028e024b381dc84a13a3570decb4,delivered,2018-06-20 06:58:43,2018-06-20 07:19:05,2018-06-25 08:05:00,NaN,2018-07-16 00:00:00
43834,2ebdfc4f15f23b91474edf87475f108e,29f0540231702fda0cfdee0a310f11aa,delivered,2018-07-01 17:05:11,2018-07-01 17:15:12,2018-07-03 13:57:00,NaN,2018-07-30 00:00:00
79263,e69f75a717d64fc5ecdfae42b2e8e086,cfda40ca8dd0a5d486a9635b611b398a,delivered,2018-07-01 22:05:55,2018-07-01 22:15:14,2018-07-03 13:57:00,NaN,2018-07-30 00:00:00
82868,0d3268bad9b086af767785e3f0fc0133,4f1d63d35fb7c8999853b2699f5c7649,delivered,2018-07-01 21:14:02,2018-07-01 21:29:54,2018-07-03 09:28:00,NaN,2018-07-24 00:00:00
92643,2d858f451373b04fb5c984a1cc2defaf,e08caf668d499a6d643dafd7c5cc498a,delivered,2017-05-25 23:22:43,2017-05-25 23:30:16,NaN,NaN,2017-06-23 00:00:00
97647,ab7c89dc1bf4a1ead9d6ec1ec8968a84,dd1b84a7286eb4524d52af4256c0ba24,delivered,2018-06-08 12:09:39,2018-06-08 12:36:39,2018-06-12 14:10:00,NaN,2018-06-26 00:00:00
98038,20edc82cf5400ce95e1afacc25798b31,28c37425f1127d887d7337f284080a0f,delivered,2018-06-27 16:09:12,2018-06-27 16:29:30,2018-07-03 19:26:00,NaN,2018-07-19 00:00:00


### Orders — Data Quality Findings

The Orders table contains 99,441 records and 8 columns.

- `order_id` is unique, with no duplicated order identifiers.
- No fully duplicated rows were found.
- Several order lifecycle date fields contain missing values.
- Most missing delivery dates are consistent with orders that were canceled, unavailable, shipped, processing or in earlier stages of the order lifecycle.
- However, 8 orders are marked as `delivered` while `order_delivered_customer_date` is missing. These records represent a potential data quality inconsistency.
- The date columns were initially loaded as strings and will need to be converted to datetime before calculating delivery times and delays.

In [42]:
date_columns = [
    "order_purchase_timestamp",
    "order_approved_at",
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "order_estimated_delivery_date"
]

orders[date_columns] = orders[date_columns].apply(pd.to_datetime)

In [43]:
orders.dtypes

order_id                                    str
customer_id                                 str
order_status                                str
order_purchase_timestamp         datetime64[us]
order_approved_at                datetime64[us]
order_delivered_carrier_date     datetime64[us]
order_delivered_customer_date    datetime64[us]
order_estimated_delivery_date    datetime64[us]
dtype: object

In [44]:
orders["order_status"].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

In [45]:
orders["order_status"].value_counts(normalize=True) * 100

order_status
delivered      97.020344
shipped         1.113223
canceled        0.628513
unavailable     0.612423
invoiced        0.315765
processing      0.302692
created         0.005028
approved        0.002011
Name: proportion, dtype: float64

#### Order Status Distribution

Approximately 97% of orders are marked as `delivered`, while the remaining orders are distributed across statuses such as shipped, canceled, unavailable, invoiced and processing.

Because order status may affect whether an order should be included in sales and revenue calculations, the treatment of non-delivered orders will be evaluated before defining the final business KPIs.

In [46]:
customers.shape

(99441, 5)

In [48]:
customers.info()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   customer_id               99441 non-null  str  
 1   customer_unique_id        99441 non-null  str  
 2   customer_zip_code_prefix  99441 non-null  int64
 3   customer_city             99441 non-null  str  
 4   customer_state            99441 non-null  str  
dtypes: int64(1), str(4)
memory usage: 3.8 MB


In [50]:
customers.isnull().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

In [52]:
customers.duplicated().sum()

np.int64(0)

In [53]:
customers["customer_id"].duplicated().sum()

np.int64(0)

In [54]:
customers["customer_unique_id"].duplicated().sum()

np.int64(3345)

In [55]:
customers["customer_unique_id"].nunique()

96096

In [56]:
customer_order_counts = customers["customer_unique_id"].value_counts()

customer_order_counts.head(10)

customer_unique_id
8d50f5eadf50201ccdcedfb9e2ac8455    17
3e43e6105506432c953e165fb2acf44c     9
1b6c7548a2a1f9037c1fd3ddfed95f33     7
6469f99c1f9dfae7733b25662e7f1782     7
ca77025e7201e3b30c44b472ff346268     7
47c1a3033b8b77b3ab6e109eb4d5fdf3     6
dc813062e0fc23409cd255f7f53c7074     6
63cfc61cee11cbe306bff5857d00bfe4     6
f0e310a6839dce9de1638e0fe5ab282a     6
de34b16117594161a6a89c50b289d35a     6
Name: count, dtype: int64

In [57]:
(customer_order_counts > 1).sum()

np.int64(2997)

In [58]:
repeat_customer_rate = (
    (customer_order_counts > 1).sum()
    / customers["customer_unique_id"].nunique()
    * 100
)

repeat_customer_rate

np.float64(3.1187562437562435)

### Customers — Data Quality Findings

The Customers table contains 99,441 records and no missing or fully duplicated rows.

- `customer_id` is unique and can be used to connect customers with the Orders table.
- `customer_unique_id` identifies the actual customer across different orders.
- There are 96,096 unique customers in the dataset.
- 2,997 customers appear in more than one order record.
- The preliminary repeat customer rate is approximately 3.1%.

The final repeat customer metric will be validated using order status to ensure that only relevant purchases are included.